source/mem/memory.go

```diff
@@ -0,0 +1,144 @@
+package mem
+
+import (
+	"fmt"
+	"io"
+	nurl "net/url"
+	"strings"
+	"sync"
+
+	"github.com/golang-migrate/migrate/v4/source"
+	"github.com/golang-migrate/migrate/v4/source/stub"
+)
+
+// init register to the source.Driver so it can be used easily
+func init() {
+	source.Register(schemeKey, &Memory{})
+}
+
+// Migration is an interface that holds migration info
+type Migration interface {
```

Comment:
Why use an interface instead of a struct?

Note: An interface may allow for an implementation of [Go-based migrations](https://github.com/golang-migrate/migrate/issues/15). e.g. logic encoded in the struct that implements the interface to return different migration strings based based on internal state